pkg/component/component.go

```diff
@@ -564,51 +572,110 @@ func ApplyConfig(client *occlient.Client, componentConfig config.LocalConfigInfo
+		if err != nil {
+			return err
+		}
+		localURLList := envSpecificInfo.GetURL()
+		for _, i := range ingressList.Items {
+			if !checkIfURLPresentInConfig(checkIfURLPresentInConfigParam{envinfoURL: localURLList, url: i.Name}) {
+				err = urlpkg.Delete(client, kClient, i.Name, componentConfig.GetApplication())
+				if err != nil {
+					return err
+				}
+				log.Successf("URL %s successfully deleted", i.Name)
+			}
+		}
+		return nil
+	} else {
+		urlList, err := urlpkg.ListPushed(client, componentConfig.GetName(), componentConfig.GetApplication())
+		if err != nil {
+			return err
+		}
+		localURLList := componentConfig.GetURL()
+		for _, u := range urlList.Items {
+			if !checkIfURLPresentInConfig(checkIfURLPresentInConfigParam{localURL: localURLList, url: u.Name}) {
+				err = urlpkg.Delete(client, kClient, u.Name, componentConfig.GetApplication())
+				if err != nil {
+					return err
+				}
+				log.Successf("URL %s successfully deleted", u.Name)
 			}
-			log.Successf("URL %s successfully deleted", u.Name)
 		}
+		return nil
 	}
-	return nil
 }
 
-func checkIfURLPresentInConfig(localURL []config.ConfigURL, url string) bool {
-	for _, u := range localURL {
-		if u.Name == url {
-			return true
+func checkIfURLPresentInConfig(checkIfURLPresentInConfigParam checkIfURLPresentInConfigParam) bool {
+	if experimental.IsExperimentalModeEnabled() {
+		for _, u := range checkIfURLPresentInConfigParam.envinfoURL {
+			if u.Name == checkIfURLPresentInConfigParam.url {
+				return true
+			}
+		}
+	} else {
+		for _, u := range checkIfURLPresentInConfigParam.localURL {
+			if u.Name == checkIfURLPresentInConfigParam.url {
+				return true
+			}
 		}
 	}
+
 	return false
 }
 
-// ApplyConfigCreateURL applies url config onto component
-func ApplyConfigCreateURL(client *occlient.Client, componentConfig config.LocalConfigInfo) error {
+type checkIfURLPresentInConfigParam struct {
+	localURL   []config.ConfigURL
+	envinfoURL []envinfo.EnvInfoURL
+	url        string
+}
 
-	urls := componentConfig.GetURL()
-	for _, urlo := range urls {
-		exist, err := urlpkg.Exists(client, urlo.Name, componentConfig.GetName(), componentConfig.GetApplication())
+// ApplyConfigCreateURL applies url config onto component
+func ApplyConfigCreateURL(client *occlient.Client, kClient *kclient.Client, componentConfig config.LocalConfigInfo, envSpecificInfo envinfo.EnvSpecificInfo) error {
+	if experimental.IsExperimentalModeEnabled() {
+		urls := envSpecificInfo.GetURL()
+		componentName, err := GetComponentName()
 		if err != nil {
-			return errors.Wrapf(err, "unable to check url")
+			return errors.Wrapf(err, "unable to get componentName")
 		}
-		if exist {
-			log.Successf("URL %s already exists", urlo.Name)
-		} else {
-			host, err := urlpkg.Create(client, urlo.Name, urlo.Port, urlo.Secure, componentConfig.GetName(), componentConfig.GetApplication())
+		for _, urlo := range urls {
+			exist, err := urlpkg.Exists(client, kClient, urlo.Name, componentName, "")
```

Comment:
Since Exists calls ListPushed, this might end up calling list multiple times. I would suggest call ListPushed once before the function and once use the output here.

Reply:
IMO, it's better to call `Exist` to list current pushed ingresses/routes everytime before creation. I don't know if concurrent creation is possible; if it is, my concern is that.
What if two users connecting to the same cluster and same namespace, with same component (Let's say they cloned from their git repo), has same `urlName` but with different url definition in their `env.yaml`, try to create url almost at the same time? get a URL list earlier and pass into the function before the creation loop would have an issue.

Reply:
> What if two users connecting to the same cluster and same namespace, with same component (Let's say they cloned from their git repo), has same urlName but with different url definition in their env.yaml, try to create url almost at the same time?

In such a race condition, I would expect the creation of later one to fail.

> get a URL list earlier and pass into the function before the creation loop would have an issue. Since a cluster can establish multiple connections by different users, it's not guaranteed there won't be any identical urlName being created in between.

I don't think identical urlNames will be a problem as the creation of one of them would simply fail in such a case. Or am I missing something?

Plus this is a extremely rare use case scenario. And even if it occurs, the cluster should handle this race condition by itself.